src/transformStateWithClones.js

```diff
@@ -7,7 +7,50 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopy = { ...state };
+  const stateVersions = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(action.keysToRemove, stateCopy);
+        break;
+
+      case 'clear':
+        clear(stateCopy);
+        break;
+
+      default:
+        throw new
+        Error(`Please check if the action.type - ${action.type} is correct`);
+    }
+
+    stateVersions.push({ ...stateCopy });
+  }
+
+  return stateVersions;
+}
+
+function addProperties(state, propertiesToAdd) {
+  for (const property in propertiesToAdd) {
+    state[property] = propertiesToAdd[property];
+  }
+}
+
+function removeProperties(keysToRemove, state) {
```

Comment:
The order of parameters in the `removeProperties` function definition should be swapped. It should be `function removeProperties(state, keysToRemove) {`